library/alloc/src/vec.rs

```diff
@@ -2391,11 +2384,24 @@ where
 }
 
 impl<T> SpecExtend<T, IntoIter<T>> for Vec<T> {
-    fn spec_extend(&mut self, mut iterator: IntoIter<T>) {
-        unsafe {
-            self.append_elements(iterator.as_slice() as _);
+    fn spec_extend(&mut self, iterator: IntoIter<T>) {
+        // Avoid reallocation if we can use iterator's storage instead. This requires 1 memcpy and 0-1 memmove
+        // while reallocation would require 1 alloc, 1-2 memcpy, 1-2 free
+        if mem::size_of::<T>() > 0
+            && self.capacity() - self.len() < iterator.len()
+            && iterator.cap - iterator.len() >= self.len()
```

Comment:
Can this be simplified as?
```suggestion
            && iterator.len() - self.len() < iterator.cap - self.capacity()
```

Reply:
The proposed change would underflow if self is larger than iterator

Reply:
Underflow?

Reply:
This is technically still an _overflow_, but that `iterator.len() - self.len()` would panic or wrap if, say, `self.capacity() == 20_000` and `self.len() == 19_950`, and `iterator.len() == 100`.

Reply:
I meant the case where `self.capacity() > iterator.cap` . The subtraction would underflow the `usize` result and thus lead to the inequality unexpectedly evaluating to `true` which would then violate the safety constraints of `into_vec_with_uninit_prefix`

Reply:
Then how about?
```suggestion
            && iterator.len().saturating_sub(self.len()) < iterator.cap.saturating_sub(self.capacity())
```
I

Reply:
For a self with `len == 2 && cap == 2` and an iterator `len == 2 && cap == 3` that would evaluate to true and attempt to store 4 elements into an allocation of 3. 💣💥